src/test/regress/sql/multi_modifying_xacts.sql

```diff
@@ -0,0 +1,57 @@
+
+ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART 1200000;
+ALTER SEQUENCE pg_catalog.pg_dist_jobid_seq RESTART 1200000;
+
+
+-- ===================================================================
+-- test end-to-end modification functionality
+-- ===================================================================
+
+CREATE TABLE researchers (
+	id bigint NOT NULL,
+    lab_id int NOT NULL,
+	name text NOT NULL
+);
+
+SELECT master_create_distributed_table('researchers', 'lab_id', 'hash');
+SELECT master_create_worker_shards('researchers', 2, 2);
+
+-- add some data
+INSERT INTO researchers VALUES (1, 1, 'Donald Knuth');
+INSERT INTO researchers VALUES (2, 1, 'Niklaus Wirth');
+INSERT INTO researchers VALUES (3, 2, 'Tony Hoare');
+INSERT INTO researchers VALUES (4, 2, 'Kenneth Iverson');
+
+-- replace a researcher, reusing their id
+BEGIN;
+DELETE FROM researchers WHERE lab_id = 1 AND id = 2;
+INSERT INTO researchers VALUES (2, 1, 'John Backus');
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 1 AND id = 2;
+
+-- abort a modification
+BEGIN;
+DELETE FROM researchers WHERE lab_id = 1 AND id = 1;
+ABORT;
+
+SELECT name FROM researchers WHERE lab_id = 1 AND id = 1;
+
+-- creating savepoints should work...
+BEGIN;
+INSERT INTO researchers VALUES (5, 3, 'Dennis Ritchie');
+SAVEPOINT hire_thompson;
+INSERT INTO researchers VALUES (6, 3, 'Ken Thompson');
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 3 AND id = 6;
+
+-- but rollback should not
+BEGIN;
+INSERT INTO researchers VALUES (7, 4, 'Jim Gray');
+SAVEPOINT hire_engelbart;
+INSERT INTO researchers VALUES (8, 4, 'Douglas Engelbart');
+ROLLBACK TO hire_engelbart;
+COMMIT;
+
+SELECT name FROM researchers WHERE lab_id = 4;
```

Comment:
Added quite a few tests, including `SELECT` after modify, DDL, `\copy`, primary key violations, and other types of errors. Used a `CONSTRAINT TRIGGER` to raise a non-constraint violation error type on a magic value. By making trigger deferred, could also test `COMMIT` failures.